File: qt/interface3/log.cpp
#include "log.h"
#include <QDebug>

namespace onelab {

log::Messages ErrorList;
log::Messages InfoList;

log::Message::Message(const QString& aSender, const QString& aContent)
{
    Sender = aSender;
    Content  = aContent;
    Time = QDateTime::currentDateTime().toString("yyyy/MM/dd HH:mm:ss.zzz");
}

int log::printf(const char* format, ...)
{
    va_list arg;
    int done;

    va_start(arg, format);
    done = vfprintf(stdout, format, arg);
    va_end(arg);

    return done;
}

void log::err(const QString &aSender, QVariant aContent)
{
    mark(aSender, aContent, ErrorList);
}

void log::info(const QString &aSender, QVariant aContent)
{
    mark(aSender, aContent, InfoList);
}

bool log::outputInfo(QIODevice *writer)
{
    return output(writer, InfoList);
}

bool log::outputError(QIODevice *writer)
{
    return output(writer, ErrorList);
}

bool log::output(QIODevice *writer, const Messages &list)
{
    QString line;    
    if (writer)
    {
        foreach(Message msg, list)
        {
            line = msg.Time + "\t" + msg.Sender + "\t" + msg.Content;            
            if (writer->write(line.toUtf8()) < 0)
                return false;
        }        
    }
    else
    {
        foreach(Message msg, list)
        {
            line = msg.Time + "\t" + msg.Sender + "\t" + msg.Content;            
            printf("%s\n", line.toStdString().c_str());
        }    
    }
    return true;    
}

void log::mark(const QString& aSender, QVariant aContent, Messages& list)
{
    if (aSender.isEmpty() || aContent.isNull()) return;
    QString content = aContent.toString();
    if (content.isEmpty()) return;
    Message msg(aSender, content);
    list << msg;
}

}

File: cs/types0/log.cs
using System;
using System.Collections.Generic;
using Messages = System.Collections.Generic.List<onelab.log.Message>;

namespace onelab
{
    public class log
    {
        public static void printf(String format, params Object[] args)
        {
            System.Console.Write(format, args);
        }

        public class Message
        {
            public String Time = DateTime.Now.ToString("yyyy/MM/dd HH:mm:ss.fff");
            public String Sender;
            public String Content;
            public Message(String aSender, String aContent)
            {
                Sender = aSender;
                Content = aContent;
            }
        }

        private static Messages ErrorList = new Messages();

        private static Messages InfoList = new Messages();

        private static void mark(String aSender, object aContent, ref Messages list)
        {
            if (string.IsNullOrWhiteSpace(aSender) || aContent == null) return;
            String content = aContent.ToString();
            if (string.IsNullOrWhiteSpace(content)) return;
            Message msg = new Message(aSender, content);
            list.Add(msg);
        }

        public static void err(String aSender, object aContent)
        {
            mark(aSender, aContent, ref ErrorList);
        }

        public static void info(String aSender, object aContent)
        {
            mark(aSender, aContent, ref InfoList);
        }

        private static bool output(System.IO.TextWriter writer, ref List<Message> list)
        {
            try
            {
                if (writer == null)
                    foreach (Message m in list)
                        System.Console.WriteLine("{0}\t{1}\t{2}", m.Time, m.Sender, m.Content);
                else
                {
                    foreach (Message m in list)
                        writer.WriteLine("{0}\t{1}\t{2}", m.Time, m.Sender, m.Content);
                    writer.Flush();
                }
            }
            catch (Exception)
            {
                return false;
            }
            return true;
        }

        public static bool outputInfo(System.IO.TextWriter writer)
        {
            return output(writer, ref InfoList);
        }

        public static bool outputError(System.IO.TextWriter writer)
        {
            return output(writer, ref ErrorList);
        }
    }
}
